kernel/kernel_drivers.go

```diff
@@ -0,0 +1,171 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package kernel
+
+import (
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"regexp"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/snap"
+)
+
+// For testing purposes
+var osSymlink = os.Symlink
+
+// We expect as a minimum something that starts with three numbers
+// separated by dots for the kernel version.
+var utsRelease = regexp.MustCompile(`^([0-9]+\.){2}[0-9]+`)
+
+// KernelVersionFromModulesDir returns the kernel version for a mounted kernel
+// snap (this would be the output if "uname -r" for a running kernel). It
+// assumes that there is a folder named modules/$(uname -r) inside the snap.
```

Comment:
I think it'd be great to introduce a set of checks that could be run at `snap pack` time to make sure that if there are kernel components defined in snap.yaml,  `modules/<kver-like-entry>/{kernel,modules*}` entries are pesent in the snap.

Reply:
Wouldn't the fact that there is a component be declared in the snap though?

Reply:
Discussed on MM, will likely need a check when the component ships modules.